### PR TITLE
JSON importer

### PR DESCRIPTION
**Pull request by d-michail**
A JSON importer to read graphs from our JSON exporter.

----
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

File: jgrapht-io/src/main/java/org/jgrapht/io/JSONImporter.java [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+                return Long.valueOf(tn.getText(), 10).toString();
+            } catch (NumberFormatException e) {
+            }
+            
+            throw new IllegalArgumentException("Failed to read valid identifier");
+        }
+
+    }
+
+    private class Node
```
Comment (jsichi) [inline jgrapht-io/src/main/java/org/jgrapht/io/JSONImporter.java]: static

File: jgrapht-io/src/main/java/org/jgrapht/io/JSONImporter.java [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+    {
+        Map<String, Attribute> attributes;
+
+        public Node(Map<String, Attribute> attributes)
+        {
+            this.attributes = attributes;
+        }
+    }
+
+    private class PartialEdge
```
Comment (jsichi) [inline jgrapht-io/src/main/java/org/jgrapht/io/JSONImporter.java]: static

File: jgrapht-io/src/test/java/org/jgrapht/io/JSONImporterTest.java [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+import java.io.*;
+
+import static org.junit.Assert.*;
+
+/**
+ * Tests for {@link JsonImporter}.
+ * 
+ * @author Dimitrios Michail
+ */
+public class JSONImporterTest
```
[the remark this answers is not in the document]
Comment (d-michail) [reply jgrapht-io/src/test/java/org/jgrapht/io/JSONImporterTest.java]: done

File: jgrapht-io/src/main/java/org/jgrapht/io/JSONImporter.java [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+                V vertex = vertexProvider.buildVertex(id, n.attributes);
+                map.put(id, vertex);
+                graph.addVertex(vertex);
+            }
+
+            // add singleton nodes
+            for (Node n : singletons) {
+                graph
+                    .addVertex(
+                        vertexProvider.buildVertex(UUID.randomUUID().toString(), n.attributes));
```
[the remark this answers is not in the document]
Comment (d-michail) [reply jgrapht-io/src/main/java/org/jgrapht/io/JSONImporter.java]: Change to using the SupplierUtil. Did not go all the way as to allow for an anonymous id provider.

**Comment by jsichi** [comment]
Thanks!

**Comment by d-michail** [comment]
Closes #768.

**Pull request merged**